Use protobuf reflection for loading runtime configuration

Molter73 opened the pull request:
## Description

This change is based upon a toy project I have at https://github.com/Molter73/config-much.

The basic idea is to use the introspection and reflection APIs from protobuf to populate the runtime configuration structures, which should allow us to add new fields to our protobuf definitions and have them magically show up in our code to use.

This is an alternative implementation to #1993.

## Checklist
- [x] Investigated and inspected CI test results
- [ ] Updated documentation accordingly

**Automated testing**
  - [ ] Added unit tests
  - [ ] Added integration tests
  - [ ] Added regression tests

If any of these don't apply, please comment below.

## Testing Performed

CI should be enough to ensure no behavioral changes have been introduced.

JoukoVirtanen commented on collector/collector.cpp:
```diff
@@ -140,8 +140,12 @@ int main(int argc, char** argv) {
   config.InitCollectorConfig(args);
-  if (!ConfigLoader::LoadConfiguration(config)) {
-    CLOG(FATAL) << "Unable to parse configuration file";
+  switch (ConfigLoader(config).LoadConfiguration()) {
```
```suggestion
  if (ConfigLoader(config).LoadConfiguration()== collecotr::ConfigLoader::PARSE_ERROR) {
```

Molter73 commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -13,109 +19,513 @@ enum PathTags {
+  // mapping for repeated fields:
+  // https://protobuf.dev/reference/cpp/api-docs/google.protobuf.message/#Reflection.GetRepeatedFieldRef.details
+  switch (field->cpp_type()) {
+    case FieldDescriptor::CPPTYPE_INT32:
+      return ParseArrayInner<int32>(msg, node, field);
+    case FieldDescriptor::CPPTYPE_UINT32:
+      return ParseArrayInner<uint32_t>(msg, node, field);
+    case google::protobuf::FieldDescriptor::CPPTYPE_INT64:
+      return ParseArrayInner<int64_t>(msg, node, field);
+    case google::protobuf::FieldDescriptor::CPPTYPE_UINT64:
+      return ParseArrayInner<uint64_t>(msg, node, field);
+    case google::protobuf::FieldDescriptor::CPPTYPE_DOUBLE:
+      return ParseArrayInner<double>(msg, node, field);
+    case google::protobuf::FieldDescriptor::CPPTYPE_FLOAT:
+      return ParseArrayInner<float>(msg, node, field);
+    case google::protobuf::FieldDescriptor::CPPTYPE_BOOL:
+      return ParseArrayInner<bool>(msg, node, field);
+    case google::protobuf::FieldDescriptor::CPPTYPE_ENUM:
+      return ParseArrayEnum(msg, node, field);
+    case google::protobuf::FieldDescriptor::CPPTYPE_STRING:
+      return ParseArrayInner<std::string>(msg, node, field);
+    case google::protobuf::FieldDescriptor::CPPTYPE_MESSAGE: {
+      return {{"Unsupport repeated type MESSAGE"}};
+    } break;
```
I'm also going to space them out a bit, it's a little hard to read otherwise.

JoukoVirtanen commented:
Thanks for doing this. This is a lot of work and you have added good unit tests, but I feel that given the amount of new code, the number of unit tests is insufficient. The unit tests call `Parse`, but there are many new functions that are introduced. I would have a lot more confidence if more of these new functions were directly tested. For example `FindUnkownFields` (Fix the typo), is not tested and none of the new unit tests, have unknown fields that are checked for.

JoukoVirtanen commented on collector/lib/ConfigLoader.h:
```diff
@@ -1,14 +1,167 @@
+   * @param node The YAML::Node to be walked.
+   * @returns an optional vector of parser errors.
+   */
+  ParserResult FindUnkownFields(const google::protobuf::Message& msg, const YAML::Node& node);
```
```suggestion
  ParserResult FindUnknownFields(const google::protobuf::Message& msg, const YAML::Node& node);
```

JoukoVirtanen replied:
I make the same typo all the time.

JoukoVirtanen commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -8,114 +14,520 @@ namespace collector {
+    }
+  }
+
+  auto res = FindUnkownFields(*msg, node);
```
Do we want to treat unknown fields in the same way as other errors. What if a user reverts to an older version, but keeps the same ConfigMap? If the new version introduced new fields, this would now cause an error.

JoukoVirtanen replied:
It might be better just to warn about unknown fields.

Molter73 replied:
This would require `ParserYaml` to have a notion of warning vs. error, it would require it to return something like a `std::pair<ParserError, ParserError>` or implement a `ParserError` with an error level. While this is achievable, it might be overkill and it will require some more changes and complexity to how `ParserYaml` itself is configured. Are you OK to keep things as are for now and we can tackle it in a follow up?

Side note, I have implemented a method to make errors ignorable in config-much itself, but I would prefer to keep us from escalating the scope of this PR even further.

- https://github.com/Molter73/config-much/commit/80beeb90391a1d7658f9ad16eae6a65b3130c6e2

JoukoVirtanen replied:
I think it could be done in a follow up PR, but I have a strong desire for that to be done before 4.8. Can you write the ticket? Thanks.

Molter73 replied:
I ended up adding the code to make it so unknown fields are ignored.

JoukoVirtanen commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -8,114 +14,520 @@ namespace collector {
+    case YAML::NodeType::Map:
+      return "Map";
+  }
+  return "";  // Unreachable
```
```suggestion
  return "Unknown"; 
```

I was able to reach your unreachable code using
```
CLOG(INFO) << "NodeTypeToString= " << NodeTypeToString(static_cast<YAML::NodeType::value>(17));
```

The output was

[INFO    2025/02/13 02:22:43] NodeTypeToString=

Molter73 replied:
This is odd, I'll have to try it, I'm fairly certain I covered all the cases for `YAML::NodeType`.

JoukoVirtanen replied:
I think I am just being a bit nit picky. I think in practice it is unlikely that part of the code would be reached, but it is possible that either a new YAML::NodeType enum is introduced or someone writes some strange code like I did. However, I prefer that the final return be something other than an empty string and I would not comment code as being unreachable unless it was inside something like `if (false)`. But then why would I have that?

Molter73 replied:
I didn't read this comment properly the first time around. This is not the way to use an enum, the whole reason enums exist is to give a limited, well defined set of values that can be used at any given time, if you are going to break that invariant then there's no code I can write that will satisfy every possible condition anyone could come up with. 

If you use the enum the way it is intended to be used, then that last line is effectively unreachable because all possible values have been accounted for in the switch-case: https://github.com/jbeder/yaml-cpp/blob/f7320141120f720aecc4c32be25586e7da9eb978/include/yaml-cpp/node/type.h#L11-L13

Now, why is the last `return` value there? Because the compiler will complain that the function reaches the end without a return statement because it is not smart enough to figure out that it is unreachable. Could I use a `default` branch in the switch case? Yes, but the advantage of doing it this way is that if you have proper linting setup for the project, (which I've wanted to implement for some time now) and you miss a value, you'll see something like this:
![image](https://github.com/user-attachments/assets/86fc0265-7894-4c7a-9cce-2f2b06cf801e)

This in turn means that if we update `yaml-cpp` in the future and an enum value is added, we would know and could act on it. If the `default` case is added though, the lint check goes away, defeating the purpose.

JoukoVirtanen commented:
In some of the existing tests you could change `ENABLED` or `DISABLED` to lower case.

Molter73 commented:
> This is a lot of work and you have added good unit tests, but I feel that given the amount of new code, the number of unit tests is insufficient.

I'm adding code coverage to the repo as we speak, the unit tests I added should cover most, if not all of the important branches.

> The unit tests call `Parse`, but there are many new functions that are introduced. I would have a lot more confidence if more of these new functions were directly tested.

What you are describing here is called white box testing and, while it may be useful in some cases, I prefer the black box testing approach that I took in the unit tests I added. The tests I added only call `Parse` which is the only public method the new `ParserYaml` class has, but internally, all of the added methods are called from there.

> For example `FindUnkownFields` (Fix the typo), is not tested and none of the new unit tests, have unknown fields that are checked for.

I thought I added a test for this, guess I missed it, I'll add it with my next commits.

JoukoVirtanen commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -8,114 +14,520 @@ namespace collector {
-  int64_t max_connections_per_minute = networking_node["maxConnectionsPerMinute"].as<int64_t>(CollectorConfig::kMaxConnectionsPerMinute);
 
+  if (errors) {
+    CLOG(ERROR) << "Failed to parse " << parser_.GetFile();
```
Since collector has multiple threads it is possible that this line will be logged, then other unrelated logging statements will be created, and then the errors in the loop will be logged, again possibly interspersed by unrelated logging statements. I think it would be best to save the error messages to one string and then log that string.

Stringy replied:
I wonder if there's any benefit to officially supporting that in the logging library (presumably it can be done by adding a specialisation of the `<<` operator?)

Molter73 replied:
I'll look into it, the logging library is purposefully unsafe w.r.t threads because you don't want to be blocking proper operations with log messages. I'm also not a big fan of copying a bunch of strings around, so in order for me to be onboard with this, I'll need to figure out whether there's a way to do it by moving strings around, but there probably isn't.

JoukoVirtanen commented on collector/lib/ConfigLoader.h:
```diff
@@ -1,14 +1,167 @@
+  std::string msg_;
+};
+
+using ParserResult = std::optional<std::vector<ParserError>>;
```
After going through the code I still don't like this. I think it is simpler to not have optional. You have some if statements checking the optional, before inserting one list of errors into another. This check is not necessary if you just use a vector. Also I think there is only one place where `empty()` is needed if optional is not used. I created a branch where I made this change and the other changes that need to be made to get everything to work with just the plain vector. Please take a look and merge it into your branch if you like it. https://github.com/stackrox/collector/pull/2042

Molter73 replied:
Answered in the PR.

TL;DR the way used in this PR is more idiomatic C++17

Stringy commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -8,114 +14,520 @@ namespace collector {
+                               const google::protobuf::FieldDescriptor* field) {
+  using namespace google::protobuf;
+
+  std::string camel;
+  const std::string* name = &field->name();
+  if (read_camelcase_) {
+    camel = SnakeCaseToCamel(*name);
+    name = &camel;
+  }
```
Pretty sure the pointer is to avoid unneeded string copies

JoukoVirtanen replied:
It seems like overkill when this won't get called very often.

JoukoVirtanen replied:
Keep it the way it is.

JoukoVirtanen commented:
> I'm adding code coverage to the repo as we speak, the unit tests I added should cover most, if not all of the important branches.

I approve.

> The tests I added only call Parse which is the only public method the new ParserYaml class has, but internally, all of the added methods are called from there.

I realize the other methods will be called as a result of calling Parse, but many of them are complicated enough that they deserve tests of their own.

Molter73 commented:
> I realize the other methods will be called as a result of calling Parse, but many of them are complicated enough that they deserve tests of their own.

I wholeheartedly disagree, most of the methods might seem complicated but are actually quite straightforward if you read them in isolation, so please provide concrete examples so we can discuss them properly.

In the end, if most of the code is tested and the output of the parsing operation is correct, what would be the benefit of testing the internals of the class? The only thing that would arise from adding such tests in my experience is that if we decide we want to change the internal implementation in the future a bunch of tests will need to be rewritten to fit, adding a lot of additional work for no added value.

JoukoVirtanen commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -8,114 +14,520 @@ namespace collector {
+}
+
+std::string ParserYaml::CamelCaseToSnake(const std::string& s) {
+  std::string out;
```
```suggestion
  std::string out;
  out.reserve(s.size());
```
Might avoid reallocation in some cases.

JoukoVirtanen commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -8,114 +14,520 @@ namespace collector {
+}
+
+std::string ParserYaml::SnakeCaseToCamel(const std::string& s) {
+  std::string out;
```
```suggestion
  std::string out;
  out.reserve(s.size());
```
Might avoid reallocation in some cases.

JoukoVirtanen commented on collector/lib/ConfigLoader.cpp:
```diff
@@ -8,114 +14,520 @@ namespace collector {
+  bool first = true;
+
+  for (const auto& c : s) {
+    if (!first && std::isupper(c) != 0) {
```
```suggestion
    if (!first && std::isupper(c)) {
```

JoukoVirtanen commented:
This is good work, and while I disagree with some things, I don't think they should be blockers to merging this. I have not seen any logical errors, and tests have been added beyond the ones that were originally created to test the feature. I added a couple of small suggestions, that are not needed. I can approve without those changes. The one thing I would like to see before approving, is to change the capitalization of the enum in some of the existing tests.

JoukoVirtanen reviewed (approved):
LGTM!
